In [21]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 4: Secure Container ---</h2><p>You arrive at the Venus fuel depot only to discover it's protected by a password.  The Elves had written the password on a sticky note, but someone <span title="Look on the bright side - isn't it more secure if nobody knows the password?">threw it out</span>.</p>
<p>However, they do remember a few key facts about the password:</p>
<ul>
<li>It is a six-digit number.</li>
<li>The value is within the range given in your puzzle input.</li>
<li>Two adjacent digits are the same (like <code>22</code> in <code>1<em>22</em>345</code>).</li>
<li>Going from left to right, the digits <em>never decrease</em>; they only ever increase or stay the same (like <code>111123</code> or <code>135679</code>).</li>
</ul>
<p>Other than the range rule, the following are true:</p>
<ul>
<li><code>111111</code> meets these criteria (double <code>11</code>, never decreases).</li>
<li><code>2234<em>50</em></code> does not meet these criteria (decreasing pair of digits <code>50</code>).</li>
<li><code>123789</code> does not meet these criteria (no double).</li>
</ul>
<p><em>How many different passwords</em> within the range given in your puzzle input meet these criteria?</p>
</article>


In [22]:
tests = [
    {
        "name": "Example 1",
        "password": 111111,
        "expected": True,
    },
    {
        "name": "Example 2",
        "password": 223450,
        "expected": False,
    },
    {
        "name": "Example 3",
        "password": 123789,
        "expected": False,
    },
]


def validate(password: int) -> bool:
    # It is a six-digit number.
    pws = str(password)
    if len(pws) != 6:
        return False
    two_adjacent_same = False
    # Two adjacent digits are the same (like 22 in 122345).
    # Going from left to right, the digits never decrease;
    # they only ever increase or stay the same (like 111123 or 135679)
    for i in range(1, len(pws)):
        two_adjacent_same |= pws[i - 1] == pws[i]
        if pws[i - 1] > pws[i]:
            return False
    return two_adjacent_same


run_tests_params(validate, tests)


Test Example 1 passed, for validate.
Test Example 2 passed, for validate.
Test Example 3 passed, for validate.
Success


In [23]:
answer = sum(1 for pw in range(272091, 815432 + 1) if validate(pw))
print(f"Part I: {answer}")

Part I: 931


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>931</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>An Elf just remembered one more important detail: the two adjacent matching digits <em>are not part of a larger group of matching digits</em>.</p>
<p>Given this additional criterion, but still ignoring the range rule, the following are now true:</p>
<ul>
<li><code>112233</code> meets these criteria because the digits never decrease and all repeated digits are exactly two digits long.</li>
<li><code>123<em>444</em></code> no longer meets the criteria (the repeated <code>44</code> is part of a larger group of <code>444</code>).</li>
<li><code>111122</code> meets the criteria (even though <code>1</code> is repeated more than twice, it still contains a double <code>22</code>).</li>
</ul>
<p><em>How many different passwords</em> within the range given in your puzzle input meet all of the criteria?</p>
</article>

</main>


In [24]:
tests = [
    {
        "name": "Example 1",
        "password": 112233,
        "expected": True,
    },
    {
        "name": "Example 2",
        "password": 123444,
        "expected": False,
    },
    {
        "name": "Example 3",
        "password": 111122,
        "expected": True,
    },
]


def validate_II(password: int) -> bool:
    # It is a six-digit number.
    pws = str(password)
    if len(pws) != 6:
        return False
    # Two adjacent digits are the same (like 22 in 122345).
    # The two adjacent matching digits are not part of a larger
    # group of matching digits
    # Going from left to right, the digits never decrease;
    # they only ever increase or stay the same (like 111123 or 135679)
    two_adjacent_same = False
    count = 1
    prev = pws[0]
    for i in range(1, len(pws)):
        if prev == pws[i]:
            count += 1
        else:
            two_adjacent_same |= count == 2
            count = 1
            prev = pws[i]

        if pws[i - 1] > pws[i]:
            return False
    return two_adjacent_same or count == 2


run_tests_params(validate_II, tests)


Test Example 1 passed, for validate_II.
Test Example 2 passed, for validate_II.
Test Example 3 passed, for validate_II.
Success


In [25]:
answer = sum(1 for pw in range(272091, 815432 + 1) if validate_II(pw))
print(f"Part II: {answer}")

Part II: 609


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>609</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
